pkg/tiller/server.go

```diff
@@ -93,3 +159,173 @@ func checkClientVersion(ctx context.Context) error {
+
+	clientset, err := sysCli.ClientSet()
+	if err != nil {
+		return c, err
+	}
+
+	// verify token
+	tokenReq := &authenticationapi.TokenReview{
+		Spec: authenticationapi.TokenReviewSpec{
+			Token: token,
+		},
+	}
+	result, err := clientset.Authentication().TokenReviews().Create(tokenReq)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return c, err
+	} else if err == nil && !result.Status.Authenticated {
+		return c, errors.New("Not authenticated")
+	}
+
+	syscfg, err := sysCli.ClientConfig()
+	if err != nil {
+		return c, err
+	}
+	usrcfg := &rest.Config{
+		Host:        syscfg.Host,
+		APIPath:     syscfg.APIPath,
+		Prefix:      syscfg.Prefix,
+		BearerToken: token,
+	}
+	usrcfg.TLSClientConfig.CertData = syscfg.TLSClientConfig.CertData
+
+	c = context.WithValue(c, kube.UserInfo, &result.Status.User)
+	c = context.WithValue(c, kube.UserClient, kube.New(&wrapClientConfig{cfg: usrcfg}))
+	c = context.WithValue(c, kube.SystemClient, sysCli)
+	return c, nil
+}
+
+func checkBasicAuth(c context.Context, h string, sysCli *kube.Client) (context.Context, error) {
+	basicAuth, err := base64.StdEncoding.DecodeString(h[len("Basic "):])
+	if err != nil {
+		return c, err
+	}
+	username, password := getUserPasswordFromBasicAuth(string(basicAuth))
+	if len(username) == 0 || len(password) == 0 {
+		return c, errors.New("Missing username or password.")
+	}
+
+	syscfg, err := sysCli.ClientConfig()
+	if err != nil {
+		return c, err
+	}
+	usrcfg := &rest.Config{
+		Host:     syscfg.Host,
+		APIPath:  syscfg.APIPath,
+		Prefix:   syscfg.Prefix,
+		Username: username,
+		Password: password,
+	}
+	usrcfg.TLSClientConfig.CertData = syscfg.TLSClientConfig.CertData
+
+	usrClient := kube.New(&wrapClientConfig{cfg: usrcfg})
+	clientset, err := usrClient.ClientSet()
+	if err != nil {
+		return c, err
+	}
+
+	// verify credentials
+	_, err = clientset.Discovery().ServerVersion()
+	if err != nil {
+		return c, err
+	}
+
+	c = context.WithValue(c, kube.UserInfo, &authenticationapi.UserInfo{
+		Username: username,
+	})
+	c = context.WithValue(c, kube.UserClient, usrClient)
+	c = context.WithValue(c, kube.SystemClient, sysCli)
+	return c, nil
+}
+
+func getUserPasswordFromBasicAuth(token string) (string, string) {
+	st := strings.SplitN(token, ":", 2)
+	if len(st) == 2 {
+		return st[0], st[1]
+	}
+	return "", ""
+}
+
+func checkClientCert(c context.Context, sysCli *kube.Client) (context.Context, error) {
+	// ref: https://github.com/grpc/grpc-go/issues/111#issuecomment-275820771
+	peer, ok := peer.FromContext(c)
+	if !ok {
+		return c, errors.New("No peer found!")
+	}
+	tlsInfo, ok := peer.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return c, errors.New("No TLS credential found!")
+	}
+	if len(tlsInfo.State.VerifiedChains) == 0 || len(tlsInfo.State.VerifiedChains[0]) == 0 {
+		return c, errors.New("No verified client certificate found!")
+	}
+
+	crt := tlsInfo.State.VerifiedChains[0][0]
+	user := authenticationapi.UserInfo{
+		Username: crt.Subject.CommonName,
+	}
```

Comment:
- [ ] What about using [`crt.Subject.Organization`](https://golang.org/pkg/crypto/x509/pkix/#Name) here to populate the `UserInfo` "Groups" field?